File: simulate_python/test/test_camera_follow.py
import math
from pathlib import Path
import sys

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from camera_follow import CameraFollower


def test_translation_yaw_smoothing_and_reset():
    q_yaw_90 = (math.sqrt(0.5), 0.0, 0.0, math.sqrt(0.5))
    follower = CameraFollower(offset=(1.0, 0.0, 0.2), azimuth=90.0)
    lookat, azimuth = follower.update((2.0, 3.0, 0.8), q_yaw_90, 0.02)
    np.testing.assert_allclose(lookat, (3.0, 3.0, 1.0), atol=1e-12)
    assert azimuth == 90.0

    follower = CameraFollower(offset=(1.0, 0.0, 0.2), azimuth=90.0, follow_yaw=True)
    lookat, azimuth = follower.update((2.0, 3.0, 0.8), q_yaw_90, 0.02)
    np.testing.assert_allclose(lookat, (2.0, 4.0, 1.0), atol=1e-12)
    assert abs(azimuth - 180.0) < 1e-12

    follower = CameraFollower(smoothing_tau=1.0)
    follower.update((0.0, 0.0, 0.0), (1.0, 0.0, 0.0, 0.0), 0.1)
    lookat, _ = follower.update((10.0, 0.0, 0.0), (1.0, 0.0, 0.0, 0.0), 0.1)
    assert 0.0 < lookat[0] < 10.0
    follower.reset()
    lookat, _ = follower.update((10.0, 0.0, 0.0), (1.0, 0.0, 0.0, 0.0), 0.1)
    assert lookat[0] == 10.0

File: simulate/test_camera_follow.cc
#include "src/camera_follow.h"

#include <cassert>
#include <cmath>

int main() {
  camera_follow::Filter filter;
  camera_follow::Settings settings;
  settings.lookat_offset = {1.0, 0.0, 0.2};
  camera_follow::Pose pose{{2.0, 3.0, 0.8}, {std::sqrt(0.5), 0.0, 0.0, std::sqrt(0.5)}};

  auto fixed = filter.Update(pose, 0.02, settings);
  assert(std::abs(fixed.lookat[0] - 3.0) < 1e-12);
  assert(std::abs(fixed.lookat[1] - 3.0) < 1e-12);
  assert(std::abs(fixed.azimuth - 90.0) < 1e-12);

  filter.Reset();
  settings.follow_yaw = true;
  auto yaw = filter.Update(pose, 0.02, settings);
  assert(std::abs(yaw.lookat[0] - 2.0) < 1e-12);
  assert(std::abs(yaw.lookat[1] - 4.0) < 1e-12);
  assert(std::abs(yaw.azimuth - 180.0) < 1e-12);

  settings.follow_yaw = false;
  settings.smoothing_tau = 1.0;
  filter.Reset();
  pose.position = {0.0, 0.0, 0.0};
  filter.Update(pose, 0.1, settings);
  pose.position = {10.0, 0.0, 0.0};
  auto smooth = filter.Update(pose, 0.1, settings);
  assert(smooth.lookat[0] > 1.0 && smooth.lookat[0] < 11.0);
  filter.Reset();
  auto reset = filter.Update(pose, 0.1, settings);
  assert(std::abs(reset.lookat[0] - 11.0) < 1e-12);
}
